lib/src/story/progress.rs

```diff
@@ -724,134 +497,7 @@ impl Story {
-    }
-
-    fn calculate_newline_output_state_change(
-        prev_text: &str,
-        curr_text: &str,
-        prev_tag_count: i32,
-        curr_tag_count: i32,
-    ) -> OutputStateChange {
-        // Simple case: nothing's changed, and we still have a newline
-        // at the end of the current content
-        let newline_still_exists = curr_text.len() >= prev_text.len()
-            && !prev_text.is_empty()
-            && curr_text.chars().nth(prev_text.len() - 1) == Some('\n');
-        if prev_tag_count == curr_tag_count
-            && prev_text.len() == curr_text.len()
-            && newline_still_exists
-        {
-            return OutputStateChange::NoChange;
-        }
-
-        // Old newline has been removed, it wasn't the end of the line after all
-        if !newline_still_exists {
-            return OutputStateChange::NewlineRemoved;
-        }
-
-        // Tag added - definitely the start of a new line
-        if curr_tag_count > prev_tag_count {
-            return OutputStateChange::ExtendedBeyondNewline;
-        }
-
-        // There must be new content - check whether it's just whitespace
-        for c in curr_text.chars().skip(prev_text.len()) {
-            if c != ' ' && c != '\t' {
-                return OutputStateChange::ExtendedBeyondNewline;
-            }
-        }
-
-        // There's new text but it's just spaces and tabs, so there's still the
-        // potential
-        // for glue to kill the newline.
-        OutputStateChange::NoChange
-    }
-
-    fn state_snapshot(&mut self) {
-        // tmp_state contains the new state and current state is stored in snapshot
-        let mut tmp_state = self.state.copy_and_start_patching();
-        std::mem::swap(&mut tmp_state, &mut self.state);
-        self.state_snapshot_at_last_new_line = Some(tmp_state);
-    }
-
-    fn discard_snapshot(&mut self) {
-        // Normally we want to integrate the patch
-        // into the main global/counts dictionaries.
-        // However, if we're in the middle of async
-        // saving, we simply stay in a "patching" state,
-        // albeit with the newer cloned patch.
-
-        if !self.async_saving {
-            self.get_state_mut().apply_any_patch();
-        }
-
-        // No longer need the snapshot.
-        self.state_snapshot_at_last_new_line = None;
-    }
-
-    fn visit_container(&mut self, container: &Rc<Container>, at_start: bool) {
-        if !container.counting_at_start_only || at_start {
-            if container.visits_should_be_counted {
-                self.get_state_mut()
-                    .increment_visit_count_for_container(container);
-            }
-
-            if container.turn_index_should_be_counted {
-                self.get_state_mut()
-                    .record_turn_index_visit_to_container(container);
-            }
-        }
-    }
-
-    fn perform_logic_and_flow_control(
+    pub(crate) fn perform_logic_and_flow_control(
```

Comment:
This method is pretty big, maybe we could extract it to `logic_and_flow_control.rs`?